### PR TITLE
Use haproxy to load balance Kube API requests

### PR DESCRIPTION
**Pull request by kiall**
Now that we can have multiple masters, we need a way for the various
services and end-users to be load balanced over the set of kube-api
servers.

We install haproxy on each node, inside a docker container, configured
to load balance requests over all the cluster masters. This haproxy
is configured to listen on 0.0.0.0 on the masters, and 127.0.0.1 on
the workers.

This is to allow the minions to simply "talk" to 127.0.0.0, and be
routed to an active kube-api server.

### CODE REVIEW

**Comment by kiall** [comment]
Most of this code was grabbed from an older change in the "ha" branch, and reworked as the diff was too large to just cherry-pick or rebase.

https://github.com/kubic-project/salt/commit/35da1bedf1116a6544693bdd5a24ce4a5eceb7ee

File: salt/haproxy/init.sls [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    - template: jinja
+    - user: root
+    - group: root
+    - mode: 644
+    - makedirs: True
+    - dir_mode: 755
+{% if "kube-master" in salt['grains.get']('roles', []) %}
+  iptables.append:
+{% else %}
+  iptables.delete:
```
Comment (robdaemon) [inline salt/haproxy/init.sls]: Why delete this? It shouldn't matter what node gets the request, they can all forward to the API server.

Comment (kiall) [inline salt/haproxy/init.sls]: While it's probably harmless to do so, we're not listening on 6443 on the workers, which would leave an open and unbound port - which I don't like the idea of :)

We could just listen on 6443 everywhere, but I think it's easier to ask people to aim the public FQDNs that compose the core service APIs to the reasonably fixed pool or masters, rather than the hopefully fluid pool of workers.

I don't have a strong opinion either way - other than -  we should only be exposing ports where we actually have a service listening on that machine

Comment (robdaemon) [inline salt/haproxy/init.sls]: Ahhh. The initial implementation I did of this on the ha branch listened on 6443 on all nodes, and forwarded to the open API servers. The local kubelets / kube-proxy all talked to localhost:6443.

In the case of any software load balancer deployed inside a cluster, and with how kube-proxy works, you are going to be listening on nodes that do not host the service and performing a reverse proxy to the appropriate place, so....

Comment (kiall) [inline salt/haproxy/init.sls]: Preface - I think you just spotted an issue, I'll need to double check in the morning.

Oh, let me rephrase slightly. We do listen on 6443 on all machines - however on masters, it's 0.0.0.0:6443, and on workers it's 127.0.0.1:6443.

The actual kube-api process listens on 0.0.0.0:6444 on the masters, and is not bound at all on the workers.

The issue I think I have is - I'm only adding 1 allow rule, when I need two (and eventually, we get network segregation, and the second allow rule is on the "internal" network)

Comment (kiall) [inline salt/haproxy/init.sls]: Actually, I add the second rule in the kube-api role, nevermind :)

File: salt/haproxy/init.sls [hunk 2]
```diff
@@ -0,0 +1,48 @@
+---
```
Comment (ereslibre) [inline salt/haproxy/init.sls]: This line can be removed.

File: salt/haproxy/init.sls [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      - {{ pillar['api']['ssl_port'] }}
+    - proto:      tcp
+
+# Send a HUP to haproxy when the config changes
+# TODO: There should be a better way to handle this, but currently, there is not. See
+# kubernetes/kubernetes#24957
+haproxy_restart:
+  cmd.run:
+    - name: |-
+        haproxy_id=$(docker ps -a | grep -E "k8s_haproxy.*\.infra\.caasp\.local_kube-system_" | awk '{print $1}')
```
Comment (ereslibre) [inline salt/haproxy/init.sls]: Minor: we only care about running containers so we should use `docker ps` instead of `docker ps -a`.

File: salt/haproxy/haproxy.cfg.jinja [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        mode tcp
+        default-server inter 10s fall 3
+        balance roundrobin
+
+{%- for minion_id, addrlist in salt['mine.get']('roles:kube-master', 'network.ip_addrs', 'grain').items() %}
+{% if addrlist is string -%}
+{% set addr = addrlist -%}
+{% else -%}
+{% set addr = addrlist|first -%}
+{% endif -%}
```
Comment (ereslibre) [inline salt/haproxy/haproxy.cfg.jinja]: What purpose has this whole block? `addr` is not used after the declaration.

Comment (kiall) [inline salt/haproxy/haproxy.cfg.jinja]: Ah, a leftover from the original commit to another branch a few months back. I'll remove it.

File: salt/haproxy/init.sls (hunk 3 above)
Comment (kiall) [inline salt/haproxy/init.sls]: Good catch.

**Comment by grahamhayes** [comment]
Approving as the issue is with the devenv tooling (using an IP in place of a FQDN)
